host/windows/syscall.c

```diff
@@ -480,142 +948,469 @@ ssize_t oe_syscall_writev_ocall(
+    if (pdir->dir_offs == 0)
+    {
+        entry->d_off = pdir->dir_offs++;
+        entry->d_type = OE_DT_DIR;
+        entry->d_reclen = sizeof(struct oe_dirent);
+        entry->d_name[0] = '.';
+        entry->d_name[1] = '\0';
+        return 0;
+    }
+
+    if (!FindNextFileW(pdir->hFind, &pdir->FindFileData))
+    {
+        DWORD winerr = GetLastError();
+
+        if (winerr == ERROR_NO_MORE_FILES)
+        {
+            /* Return 1 to indicate there no more entries. */
+            return 1;
+        }
+        else
+        {
+            _set_errno(_winerr_to_errno(winerr));
+            return -1;
+        }
+    }
+
+    nlen = WideCharToMultiByte(
+        CP_UTF8, 0, pdir->FindFileData.cFileName, -1, NULL, 0, NULL, NULL);
+    (void)WideCharToMultiByte(
+        CP_UTF8,
+        0,
+        pdir->FindFileData.cFileName,
+        nlen,
+        entry->d_name,
+        sizeof(entry->d_name),
+        NULL,
+        NULL);
+
+    entry->d_type = 0;
+    if (pdir->FindFileData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+    {
+        entry->d_type = OE_DT_DIR;
+    }
+    else if (pdir->FindFileData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)
+    {
+        entry->d_type = OE_DT_LNK;
+    }
+    else
+    {
+        entry->d_type = OE_DT_REG;
+    }
+
+    entry->d_off = pdir->dir_offs++;
+    entry->d_reclen = sizeof(struct oe_dirent);
+
+    return 0;
 }
 
 void oe_syscall_rewinddir_ocall(uint64_t dirp)
 {
-    OE_UNUSED(dirp);
+    DWORD err = 0;
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
+    WCHAR* wpathname = pdir->pdirpath;
+    // Undo abosolute path forcing again. We do this over because we need to
+    // preserve the allocation address for free.
+    if (wpathname[0] == '/' && wpathname[2] == ':')
+    {
+        wpathname++;
+    }
 
-    PANIC;
+    FindClose(pdir->hFind);
+    memset(&pdir->FindFileData, 0, sizeof(pdir->FindFileData));
+
+    pdir->hFind = FindFirstFileW(wpathname, &pdir->FindFileData);
+    if (pdir->hFind == INVALID_HANDLE_VALUE)
+    {
+        err = GetLastError();
+    }
+    pdir->dir_offs = 0;
 }
 
 int oe_syscall_closedir_ocall(uint64_t dirp)
 {
-    OE_UNUSED(dirp);
+    struct WIN_DIR_DATA* pdir = (struct WIN_DIR_DATA*)dirp;
 
-    PANIC;
+    if (!dirp)
+    {
+        return -1;
+    }
+    if (!FindClose(pdir->hFind))
+    {
+        return -1;
+    }
+    free(pdir->pdirpath);
+    pdir->pdirpath = NULL;
+    free(pdir);
+    return 0;
 }
 
 int oe_syscall_stat_ocall(const char* pathname, struct oe_stat* buf)
 {
-    OE_UNUSED(pathname);
-    OE_UNUSED(buf);
+    int ret = -1;
+    WCHAR* wpathname = oe_syscall_path_to_win(pathname, NULL);
+    struct _stat64 winstat = {0};
 
-    PANIC;
+    ret = _wstat64(wpathname, &winstat);
+    if (ret < 0)
+    {
+        // How do we get to  wstat's error
+
+        _set_errno(_winerr_to_errno(GetLastError()));
+        goto done;
+    }
+
+#undef st_atime
```

Comment:
why is this being undefed?

Reply:
@radhikaj This is a work around here. Due to the macro here https://github.com/openenclave/openenclave/blob/e83f101b2d0113680bc10359edc05ab4b92e942e/include/openenclave/internal/syscall/sys/stat.h#L53 the winstat.st_atime will be expanded as winstat.st_atim.tv_sec.

Reply:
Will this macro over be used?  #define st_atime st_atim.tv_sec 
I am wondering if it is used, should it be redefed after?
If not, can we undef it in a more central location?

Reply:
Yes for older version POSIX device this macro is necessary. At the boundary between Windows and POSIX the conflicts arise. I think undef here is the most convenient method, which is the same method for cygwin.

Reply:
Fixed, thanks.